ssl/statem/statem_dtls.c

```diff
@@ -49,76 +49,57 @@ static int dtls_get_reassembled_message(SSL_CONNECTION *s, int *errtype,
 
 static dtls_sent_msg *dtls1_sent_msg_new(size_t msg_len)
 {
-    dtls_sent_msg *msg = NULL;
-    unsigned char *msg_buf = NULL;
+    dtls_sent_msg *msg = OPENSSL_malloc(sizeof(*msg) + msg_len);
 
-    if ((msg = OPENSSL_zalloc(sizeof(*msg))) == NULL)
+    if (msg == NULL)
         return NULL;
 
-    if (msg_len) {
-        if ((msg_buf = OPENSSL_malloc(msg_len)) == NULL) {
-            OPENSSL_free(msg);
-            return NULL;
-        }
-    }
+    memset(msg, 0, sizeof(*msg));
 
     /* zero length msg gets msg->msg_buf == NULL */
-    msg->msg_buf = msg_buf;
+    if (msg_len > 0)
+        msg->msg_buf = (unsigned char *)(msg + 1);
 
     return msg;
 }
 
 void dtls1_sent_msg_free(dtls_sent_msg *msg)
 {
-    if (msg == NULL)
-        return;
-
-    OPENSSL_free(msg->msg_buf);
-    OPENSSL_free(msg);
+    if (msg != NULL)
+        OPENSSL_free(msg);
 }
 
 static hm_fragment *dtls1_hm_fragment_new(size_t frag_len, int reassembly)
 {
-    hm_fragment *frag = NULL;
-    unsigned char *buf = NULL;
-    unsigned char *bitmask = NULL;
+    const size_t bitmask_len = (reassembly ? RSMBLY_BITMASK_SIZE(frag_len) : 0);
+    hm_fragment *frag = OPENSSL_malloc(sizeof(*frag) + frag_len + bitmask_len);
 
-    if ((frag = OPENSSL_zalloc(sizeof(*frag))) == NULL)
+    if (frag == NULL)
         return NULL;
 
-    if (frag_len) {
-        if ((buf = OPENSSL_malloc(frag_len)) == NULL) {
-            OPENSSL_free(frag);
-            return NULL;
-        }
-    }
+    memset(frag, 0, sizeof(*frag));
```

Comment:
as above, seems like zalloc would be more appropriate here

Reply:
To avoid unnecessarily clearing the buffer.